zcash_primitives/src/sapling/value.rs

```diff
@@ -0,0 +1,228 @@
+//! - Define your own "amount" type for note values, and convert it to `NoteValue` prior
+//!   to calling [`SaplingBuilder::add_output`].
+//!
+//! Inside the circuit, note values are constrained to be unsigned 64-bit integers.
+//!
+//! # Caution!
+//!
+//! An `i64` is _not_ a signed 64-bit integer! The [Rust documentation] calls `i64` the
+//! 64-bit signed integer type, which is true in the sense that its encoding in memory
+//! takes up 64 bits. Numerically, however, `i64` is a signed 63-bit integer.
+//!
+//! Fortunately, users of this crate should never need to construct [`ValueSum`] directly;
+//! you should only need to interact with [`NoteValue`] (which can be safely constructed
+//! from a `u64`) and `valueBalanceSapling` (which can be represented as an `i64`).
+//!
+//! [`Bundle`]: crate::transaction::components::sapling::Bundle
+//! [`Bundle::value_balance`]: crate::transaction::components::sapling::Bundle::value_balance
+//! [`SaplingBuilder::value_balance`]: crate::transaction::components::sapling::builder::SaplingBuilder::value_balance
+//! [`SaplingBuilder::add_output`]: crate::transaction::components::sapling::builder::SaplingBuilder::add_output
+//! [Rust documentation]: https://doc.rust-lang.org/stable/std/primitive.i64.html
+
+use ff::Field;
+use group::GroupEncoding;
+use rand::RngCore;
+use subtle::CtOption;
+
+use crate::constants::{VALUE_COMMITMENT_RANDOMNESS_GENERATOR, VALUE_COMMITMENT_VALUE_GENERATOR};
+
+mod sums;
+pub use sums::{CommitmentSum, OverflowError, TrapdoorSum, ValueSum};
+
+/// Maximum note value.
+pub const MAX_NOTE_VALUE: u64 = u64::MAX;
+
+/// The non-negative value of an individual Sapling note.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+pub struct NoteValue(u64);
+
+impl NoteValue {
+    /// Returns the raw underlying value.
+    pub fn inner(&self) -> u64 {
+        self.0
+    }
+
+    /// Creates a note value from its raw numeric value.
+    ///
+    /// This only enforces that the value is an unsigned 64-bit integer. Callers should
+    /// enforce any additional constraints on the value's valid range themselves.
+    pub fn from_raw(value: u64) -> Self {
+        NoteValue(value)
+    }
+}
+
+/// The blinding factor for a [`ValueCommitment`].
+#[derive(Clone, Debug)]
+pub struct ValueCommitTrapdoor(jubjub::Scalar);
+
+impl ValueCommitTrapdoor {
+    /// Generates a new value commitment trapdoor.
+    ///
+    /// This is public for access by `zcash_proofs`.
+    pub fn random(rng: impl RngCore) -> Self {
+        ValueCommitTrapdoor(jubjub::Scalar::random(rng))
+    }
+
+    /// Returns the inner Jubjub scalar representing this trapdoor.
+    ///
+    /// This is public for access by `zcash_proofs`.
+    pub fn inner(&self) -> jubjub::Scalar {
+        self.0
+    }
+}
+
+/// A commitment to a [`ValueSum`].
+#[derive(Clone, Debug)]
+pub struct ValueCommitment(jubjub::ExtendedPoint);
+
+impl ValueCommitment {
+    /// Derives a `ValueCommitment` by $\mathsf{ValueCommit^{Sapling}}$.
```

Comment:
If this is intended to exclude small-order points, please document that and enforce it in `derive`/`derive_inner`.